org.eclipse.xtext/src/org/eclipse/xtext/scoping/Scopes.java

```diff
@@ -86,46 +80,38 @@ public static Iterable<IEObjectDescription> scopedElementsFor(Iterable<? extends
 	 * filtered out.
 	 */
 	public static <T extends EObject> Iterable<IEObjectDescription> scopedElementsFor(Iterable<? extends T> elements,
-			final Function<T, QualifiedName> nameComputation) {
-		Iterable<IEObjectDescription> transformed = Iterables.transform(elements,
-				new Function<T, IEObjectDescription>() {
-					@Override
-					public IEObjectDescription apply(T from) {
-						final QualifiedName qualifiedName = nameComputation.apply(from);
-						if (qualifiedName != null)
-							return new EObjectDescription(qualifiedName, from, null);
-						return null;
-					}
-				});
+			Function<T, QualifiedName> nameComputation) {
+		Iterable<IEObjectDescription> transformed = Iterables.transform(elements, from -> {
+			QualifiedName qualifiedName = nameComputation.apply(from);
+			if (qualifiedName != null)
+				return new EObjectDescription(qualifiedName, from, null);
+			return null;
+		});
 		return Iterables.filter(transformed, Predicates.notNull());
 	}
-	
+
 	/**
 	 * indexes the IEObject description using the given
 	 */
 	public static <T> Multimap<T,IEObjectDescription> index(Iterable<IEObjectDescription> descriptions, Function<IEObjectDescription,T> indexer) {
-		ArrayList<IEObjectDescription> list = Lists.newArrayList(descriptions);
+		List<IEObjectDescription> list = new ArrayList<>();
```

Comment:
It's ok here, but keep in mind that `Lists.newArrayList(Iterable)` has a special casing for Collections and will call the more optimal `ArrayList(Collection)` constructor, if possible.

Reply:
Yes, that's right. But since the purpose of the entire creation of the `ArrayList` seems to be being able to determine the size of the `Iterable` and at the same time avoiding the second pass over the `Iterable` for the subsequent iteration, the best would be to to test if the `Iterable` is collection and act accordingly or provide an overload for collections.
The latter could then be called by the current method after the iterable content was copied into a List.